TYPESCRIPT.md

```diff
@@ -0,0 +1,150 @@
+    <img src="https://img.shields.io/badge/-Join%20Community%20Slack-2AB27B.svg?logo=slack&maxAge=31557600" alt="Join Contentful Community Slack">
+  </a>
+</p>
+
+<!-- shared header  END --> 
+
+
+## Table of contents
+- [Introduction](#introduction)
+- [Query types](#query-types)
+    - [Static query types](#static-query-keys)
+    - [Dynamic query types](#dynamic-field-query-keys)
+- [Response types](#response-types)
+    - [withAllLocales](#withalllocales)
+- [Generating type definitions](#generating-type-definitions-for-content-types)
+
+## Introduction
+<a href="LICENSE">
+    <img src="https://img.shields.io/badge/license-MIT-brightgreen.svg" alt="MIT License" />
+  </a><a href="https://travis-ci.org/contentful/contentful.js">
+    <img src="https://travis-ci.org/contentful/contentful.js.svg?branch=master" alt="Build Status">
+  </a>
+ <a href="https://www.npmjs.com/package/contentful">
+    <img src="https://img.shields.io/npm/v/contentful.svg" alt="NPM">
+  </a>
+   <a href="https://www.jsdelivr.com/package/npm/contentful">
+    <img src="https://data.jsdelivr.com/v1/package/npm/contentful/badge" alt="jsDelivr Hits">
+  </a>
+<a href="https://npm-stat.com/charts.html?package=contentful">
+    <img src="https://img.shields.io/npm/dm/contentful.svg" alt="NPM downloads">
+  </a>
+<a href="https://unpkg.com/contentful/dist/contentful.browser.min.js">
+    <img src="https://img.badgesize.io/https://unpkg.com/contentful/dist/contentful.browser.min.js?compression=gzip" alt="GZIP bundle size">
+  </a>
+
+With version `10.0.0` we have completely rewritten the client to give the user more support on types.
+
+
+
+## Query types
+When querying for entries and assets, you get full type support for keys and values.
+This applies to:
+- `getEntry`
+- `getEntries`
+- `getAsset`
+- `getAssets`
+
+We have 2 levels of support:
+
+### Static query keys
+![](images/static-query-keys.png)
+
+```
+getEntries({
+    'skip': 10,
+    'limit': 20,
+    'include': 5
+})
+```
+
+### Dynamic (field) query keys
+![](images/dynamic-query-keys.png)
+
+Dynamic query keys are based on the given shape of the expected entries content type.
+To calculate dynamic keys, we have to pass-in the shape of the fields:
+```typescript
+type ExampleEntryFields = {
+  productName: Contentful.EntryFields.Text,
+  image: Contentful.Asset,
+  price: Contentful.EntryFields.Number,
+  categories: Contentful.Entry<CategoryEntryFields>[],
+}
+```
+
+This gives us the relevant information needed, to calculate the dynamic keys and their possible value types.
+```
+getEntries<ExampleEntryFields>({
+    'fields.price[in]': [10, 20]
+})
+```
+
+#### Limitation
+Due to the limitation with recursive types, we can only calculate keys on the root level of your content type.
+There is currently no way to calculate keys for nested (recursive) content types.
+
+## Response types
+With version `10.0.0` we introduce chained clients to make better assumptions on response types.
```

Comment:
Maybe add a link to the explanation in the default Readme (https://github.com/contentful/contentful.js/blob/typescript/update-documentation/README.md#chained-clients)?